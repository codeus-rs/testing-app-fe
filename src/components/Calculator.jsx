import React, { useState } from 'react';
import axios from 'axios';
// const axios = require('axios');

const Calculator = () => {
  const [input, setInput] = useState('');
  const [result, setResult] = useState('');

  const handleClick = (value) => {
    setInput(input + value);
  };

  const handleClear = () => {
    setInput('');
    setResult('');
  };

  const handleCalculate = async () => {
    try {
      const response = await axios.post('http://localhost:5001/calculate', {
        expression: input,
      });
      setResult(response.data.result);
    } catch (error) {
      console.error('Error calculating:', error);
    }
  };

  return (
    <div className="container">
      <h1>Calculator</h1>
      <div className="calculator">
        <input type="text" value={input} readOnly data-testid="input-field" />
        <div className="buttons">
          {'1234567890+-*/'.split('').map((char) => (
            <button key={char} onClick={() => handleClick(char)}>
              {char}
            </button>
          ))}
          <button onClick={handleClear} data-testid="button-C">
            C
          </button>
          <button onClick={handleCalculate}>=</button>
        </div>
      </div>
      {result && <div className="result">Result: {result}</div>}
    </div>
  );
};

export default Calculator;
