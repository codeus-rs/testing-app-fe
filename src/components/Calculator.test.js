import React from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import Calculator from './Calculator';
import axios from 'axios';

// Mock the axios module
jest.mock('axios');

describe('Calculator Component', () => {
  beforeEach(() => {
    // Reset axios mock before each test
    axios.post.mockReset();
  });

  test('renders calculator with buttons and input', () => {
    render(<Calculator />);

    expect(screen.getByTestId('input-field')).toBeInTheDocument();

    expect(screen.getByText('12')).toBeInTheDocument();
    expect(screen.getByText('C')).toBeInTheDocument();
    expect(screen.getByText('=')).toBeInTheDocument();
  });

  test('updates input value when buttons are clicked', () => {
    render(<Calculator />);

    fireEvent.click(screen.getByText('1'));
    fireEvent.click(screen.getByText('2'));

    expect(screen.getByTestId('input-field')).toHaveValue('12');
  });

  test('clears input and result when "C" is clicked', () => {
    render(<Calculator />);

    fireEvent.click(screen.getByText('1'));
    fireEvent.click(screen.getByText('2'));

    fireEvent.click(screen.getByText('C'));

    expect(screen.getByTestId('input-field')).toHaveValue('');
    expect(screen.queryByText(/Result:/)).not.toBeInTheDocument();
  });

  test('calculates result and displays it', async () => {
    // Mock axios.post to return a successful response
    axios.post.mockResolvedValue({
      data: { result: '42' },
    });

    render(<Calculator />);

    fireEvent.click(screen.getByText('4'));
    fireEvent.click(screen.getByText('2'));
    fireEvent.click(screen.getByText('='));

    await waitFor(() => {
      expect(screen.getByText('Result: 42')).toBeInTheDocument();
    });
  });

  test('handles errors from the calculation API', async () => {
    // Mock axios.post to return an error
    axios.post.mockRejectedValue(new Error('Error calculating'));

    render(<Calculator />);

    fireEvent.click(screen.getByText('1'));
    fireEvent.click(screen.getByText('2'));
    fireEvent.click(screen.getByText('='));

    await waitFor(() => {
      expect(screen.queryByText(/Result:/)).not.toBeInTheDocument();
    });
  });
});

//Snapshot test
test('renders CurrencyConverter correctly', () => {
  const { asFragment } = render(<Calculator />);
  expect(asFragment()).toMatchSnapshot();
});
