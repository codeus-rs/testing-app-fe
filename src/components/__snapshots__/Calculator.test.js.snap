// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders CurrencyConverter correctly 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <h1>
      Calculator
    </h1>
    <div
      class="calculator"
    >
      <input
        data-testid="input-field"
        readonly=""
        type="text"
        value=""
      />
      <div
        class="buttons"
      >
        <button>
          1
        </button>
        <button>
          2
        </button>
        <button>
          3
        </button>
        <button>
          4
        </button>
        <button>
          5
        </button>
        <button>
          6
        </button>
        <button>
          7
        </button>
        <button>
          8
        </button>
        <button>
          9
        </button>
        <button>
          0
        </button>
        <button>
          +
        </button>
        <button>
          -
        </button>
        <button>
          *
        </button>
        <button>
          /
        </button>
        <button>
          C
        </button>
        <button>
          =
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
